enhancements/rhcos/split-rhcos-into-layers.md

```diff
@@ -0,0 +1,467 @@
+OCP layered build.
+
+#### Add more comprehensive CI
+
+Another important follow-up would to be to hook up better CI testing to openshift/os. There is
+currently no CI test on that repo which actually launches a cluster. The reason is that RHCOS
+is just built differently. But now that the node image is simply yet another layered image build,
+it fits perfectly in Prow's opinionated model of building the image and shoving it in the test
+release payload with which to launch test cluster.
+
+#### Re-organize MCO templates
+
+Currently, there's a pile of templated files that live in the MCO. These
+files configure systemd services and run extra logic as required by OCP and
+conceptually have nothing to do with the MCO itself. It would then make sense to
+move them to the openshift/os repo alongside the rest of the definition for the
+node image.
+
+In a first step, we would leave them in the MCO but work on "de-templatizing"
+them by leveraging existing systemd directives as much as possible (e.g.
+conditionalizing on platform can be done with `ignition.platform.id=azure`,
+conditionalizing on OVN could be done with a generator or by having it pulled in
+by another systemd unit, etc...).
+
+Once they're fully de-templatized, we can move them to openshift/os so that they
+are part of the node image proper.
+
+This would considerably shrink the amount of templates that the MCO would
+need to process, and as a result serve via the MCS and/or layer via on-cluster
+layering.
+
+### Risks and Mitigations
+
+#### Versioning scheme
+
+The versioning scheme for the *node image* will remain unchanged (e.g.
+418.94.202409090849-0). But the versioning scheme for the *bootimages* will
+change from e.g. 418.94.202409090849-0 to e.g. 9.4.202409090849-0 since they'll
+no longer be versioned with OpenShift. (But note that the format itself remains
+the same, and is still semver-compatible.)
+
+This change is likely to throw off anything that currently parses the version
+string to for example derive the OCP and RHEL versions for which an RHCOS image
+was built. Those will need to be adapted if they deal with bootimages. The
+release browser as mentioned above is a likely candidate.
+
+Additionally, RHCOS artifacts on the official mirrors are currently versioned
+by the OCP release (e.g. `rhcos-4.16.3-x86_64-live.x86_64.iso`) and kept under
+OCP-versioned directories (e.g. `4.16.3/`). We don't want customers to have to
+know which version of RHCOS bootimages to download for their target OCP version.
+So to avoid confusing customers, we should still have OCP-versioned directories.
+Ideally the mirror server could symlink/somehow dedupe the bootimages shared
+across OCP versions but this is of course not a requirement.
+
+As such, the versioning change should be transparent to end-users, but on the
+off chance that a customer is relying on the current semantics, we should still
+announce it in the GA release notes.
```

Comment:
And we'd start to see installer PRs moving 4.19-4.21 all to the same boot images?

Reply:
Tangentially related to this is also moving static templates from the MCO to the coreos overlay: https://issues.redhat.com/browse/MCO-1429

Reply:
> Always have the installer fetch it from the configmap in the target release payload.

At the moment (or at least as of my last installer sync, I might be stale), there's no requirement for the installer to be able to access the release image it's launching.  E.g. an installer could be talking to cloud-provider APIs to create infra, and to the Kube API via a proxied pipe, but it does not need access to the registry holding the release image.  You could maybe dance around that by grabbing boot image data and stuffing it into the installer binaries at build time, although I'm not sure how complicated that would be at the build-tooling layer.  Or you could also teach `oc adm release extract ...` to inject it at extraction time (as [it already does with some other release properties][1]).

Or maybe the installer functioning without access to the release-image registry isn't that big a deal.  I'd taken a pass at this kind of access way back in openshift/installer#1286, and it seems to have failed more on "meh, not worth the effort" than on "this adds registry-access dependencies we don't want to carry".

[1]: https://github.com/openshift/oc/blob/8e9a2be04f902d58b3be21b92ed9016854640cd6/pkg/cli/admin/release/extract_tools.go#L57-L59

Reply:
> there's no requirement for the installer to be able to access the release image it's launching
Note that ABI currently does require to access the release image during the installation for various bits. In particular, for the disconnected installations, the live ISO image is [extracted](https://github.com/openshift/installer/blob/19c87c40a42acd92323c32b62af8de605498e5bb/pkg/asset/agent/image/baseiso.go#L216) from the `machine-os-images` release image